kore/src/Kore/Step/Simplification/Predicate.hs

```diff
@@ -0,0 +1,330 @@
+
+{- | If the arguments of 'Or' are already in 'NormalForm', then normalization is
+ trivial.
+
+ @normalizeOr@ obeys these laws:
+
+ Identity:
+
+ @
+ \\or(\\bottom, P[1]) = P[1]
+ @
+
+ Annihilation:
+
+ @
+ \\or(\\top, _) = \\top
+ @
+
+ Idempotence:
+
+ @
+ \\or(P[1], P[1]) = P[1]
+ @
+-}
+normalizeOr ::
+    Applicative simplifier =>
+    Or sort NormalForm ->
+    simplifier NormalForm
+normalizeOr = pure . fold
+{-# INLINE normalizeOr #-}
+
+-- | 'Bottom' is regarded as trivially-normalizable.
+normalizeBottom ::
+    Applicative simplifier =>
+    Bottom sort NormalForm ->
+    simplifier NormalForm
+normalizeBottom _ = pure MultiOr.bottom
+{-# INLINE normalizeBottom #-}
+
+-- | 'Top' is regarded as trivially-normalizable.
+normalizeTop ::
+    Applicative simplifier =>
+    Top sort NormalForm ->
+    simplifier NormalForm
+normalizeTop _ = pure (MultiOr.singleton MultiAnd.top)
+{-# INLINE normalizeTop #-}
+
+{- | @simplifyNot@ obeys these laws:
+
+ 'Top':
+
+ @
+ \\not(\\top) = \\bottom
+ @
+
+ 'Bottom':
+
+ @
+ \\not(\\bottom) = \\top
+ @
+
+ 'Not':
+
+ @
+ \\not(\\not(P)) = P
+ @
+
+ 'Or':
+
+ @
+ \\not(\\or(P[1], P[2])) = \\and(\\not(P[1]), \\not(P[2]))
+ @
+
+ @simplifyNot@ does not expand @\not(\and(_, _))@ into @\or(_, _)@, because
+ the purpose of simplification is mostly to prepare 'Predicate' for the
+ external solver or for the user, and the un-expanded form is more compact.
+-}
+simplifyNot ::
+    forall simplifier sort.
+    Monad simplifier =>
+    Not sort NormalForm ->
+    simplifier NormalForm
+simplifyNot Not{notChild = multiOr, notSort} = do
+    disjunctiveNormalForms <- Logic.observeAllT $ do
+        multiAnd <- Logic.scatter multiOr
+        normalizeNotAnd Not{notSort, notChild = multiAnd} & lift
+    normalizeMultiAnd (MultiAnd.make disjunctiveNormalForms)
+
+normalizeNotAnd ::
+    forall simplifier sort.
+    Monad simplifier =>
+    Not sort (MultiAnd (Predicate RewritingVariableName)) ->
+    simplifier NormalForm
+normalizeNotAnd Not{notSort, notChild = predicates} =
+    case toList predicates of
+        [] ->
+            -- \not(\top)
+            bottom
+        [predicate] ->
+            case predicateF of
+                NotF Not{notChild = result} ->
+                    MultiAnd.fromPredicate result
+                        & MultiOr.singleton
+                        & pure
+                _ -> fallback
+          where
+            _ :< predicateF = Recursive.project predicate
+        _ -> fallback
+  where
+    fallback =
+        -- \not(\and(_, ...))
+        MultiAnd.toPredicate predicates
+            & fromNot
+            & Predicate.markSimplified
+            & MultiAnd.singleton
+            & MultiOr.singleton
+            & pure
+    bottom = normalizeBottom Bottom{bottomSort = notSort}
+
+{- |
+ @
+ \\implies(L, R) = \\or(\\not(L), \\and(L, R))
+ @
+
+ Note: @L@ is carried through to the right-hand side of 'Implies' to maximize
+ the information content of that branch.
```

Comment:
We simplify `\implies` sometimes as the result of `\equals` simplification, if either side is not a function-like pattern. This isn't very common. Carrying the left-hand side over improves simplification a bit in these cases; it can rule out some branches which were also carrying `\not(L)` from another part.